### PR TITLE
Clean smt2 parsing of named attributes

### LINKED ISSUE #2919: Move helper functions out of SMT2 parser
Helper functions should be moved out of the parser into the Smt2 class:
- https://github.com/CVC4/CVC4/blob/master/src/parser/smt2/Smt2.g#L153
- https://github.com/CVC4/CVC4/blob/master/src/parser/smt2/Smt2.g#L183

### PR DESCRIPTION
**Pull request by ajreynol**
Fixes #2919.

### CODE REVIEW

File: src/parser/smt2/smt2.cpp [hunk 1]
```diff
@@ -1566,5 +1566,29 @@ InputLanguage Smt2::getLanguage() const
...
+    parseError("improperly formed :named annotation");
+  }
+  std::string name = sexpr.getValue();
+  checkUserSymbol(name);
+  // ensure expr is a closed subterm
+  if (expr.hasFreeVariable())
+  {
+    std::stringstream ss;
+    ss << ":named annotations can only name terms that are closed";
+    parseError(ss.str());
```
Comment (mpreiner) [inline src/parser/smt2/smt2.cpp]: I guess there is no way to report the free variables here as it was done in the old code?

File: src/parser/smt2/smt2.h [hunk 2]
```diff
@@ -425,6 +425,17 @@ class Smt2 : public Parser
...
     }
   }
+  /** Set named attribute
+   *
+   * This is called when expression expr is annotated with a name, i.e.
+   * (! expr :named sexpr). It sets up the necessary information to process
+   * this naming, including marking that expr is the last named term.
+   *
+   * We construct an expression symbol whose name is the name of s-expression
+   * which is used later for tracking assertions in unsat cores. This
```
Comment (mpreiner) [inline src/parser/smt2/smt2.h]:
```suggestion
   * sexpr, which is used later for tracking assertions in unsat cores. This
```

The name of s-expression you are referring to is the argument `sexpr` right?

Comment (ajreynol) [reply src/parser/smt2/smt2.h]: No, it is a symbol we construct whose name is the same as the `sexpr` (at least that's how its currently implemented).

File: src/parser/smt2/smt2.cpp (hunk 1 above)
Comment (ajreynol) [reply src/parser/smt2/smt2.cpp]: Not unless we extend the ExprManager interface; this is super low priority so I don't think it was worth the trouble.

**Pull request merged**